source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -2016,4 +2048,90 @@ namespace AdaptiveCards { namespace Uwp
+        if (actionType == ABI::AdaptiveCards::Uwp::ActionType::ShowCard)
+        {
+            ComPtr<IAdaptiveActionsConfig> actionsConfig;
+            THROW_IF_FAILED(m_hostConfig->get_Actions(actionsConfig.GetAddressOf()));
+            ComPtr<IAdaptiveShowCardActionConfig> showCardActionConfig;
+            THROW_IF_FAILED(actionsConfig->get_ShowCard(&showCardActionConfig));
+            ABI::AdaptiveCards::Uwp::ActionMode showCardActionMode;
+            THROW_IF_FAILED(showCardActionConfig->get_ActionMode(&showCardActionMode));
+            if (showCardActionMode == ABI::AdaptiveCards::Uwp::ActionMode::Inline)
+            {
+                // Was inline show card, so don't wrap the element and just return
+                ComPtr<IUIElement> localElementToWrap(elementToWrap);
+                localElementToWrap.CopyTo(finalElement);
+                return;
+            }
+        }
+
+        ComPtr<IButton> button = XamlHelpers::CreateXamlClass<IButton>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_Button));
+
+        ComPtr<IContentControl> buttonAsContentControl;
+        THROW_IF_FAILED(button.As(&buttonAsContentControl));
+        THROW_IF_FAILED(buttonAsContentControl->put_Content(elementToWrap));
+
+        ComPtr<IAdaptiveSpacingConfig> spacingConfig;
+        THROW_IF_FAILED(m_hostConfig->get_Spacing(&spacingConfig));
+
+        UINT32 cardPadding;
+        THROW_IF_FAILED(spacingConfig->get_Padding(&cardPadding));
+
+        double negativeCardMargin = cardPadding * -1.0;
+
+        // TODO: Apply negative margin to top/bottom that causes button to appear halfway between spacing between elements.
+        // However this will be tricky, since to get the spacing for the bottom, we need to know the NEXT card element.
+
+        ComPtr<IControl> buttonAsControl;
+        THROW_IF_FAILED(button.As(&buttonAsControl));
+        THROW_IF_FAILED(buttonAsControl->put_HorizontalContentAlignment(HorizontalAlignment_Stretch));
+        ComPtr<IBrush> buttonBackgroundBrush = GetSolidColorBrush(Color());
+        THROW_IF_FAILED(buttonAsControl->put_Background(buttonBackgroundBrush.Get()));
+        THROW_IF_FAILED(buttonAsControl->put_Padding({ (double)cardPadding, 0, (double)cardPadding, 0 }));
+
+        ComPtr<IFrameworkElement> buttonAsFrameworkElement;
+        THROW_IF_FAILED(button.As(&buttonAsFrameworkElement));
+        THROW_IF_FAILED(buttonAsFrameworkElement->put_HorizontalAlignment(HorizontalAlignment_Stretch));
+
+        THROW_IF_FAILED(buttonAsFrameworkElement->put_Margin({ negativeCardMargin, 0, negativeCardMargin, 0 }));
+
+        WireButtonClickToAction(button.Get(), action, renderContext);
+
+        THROW_IF_FAILED(button.CopyTo(finalElement));
+    }
+
+    void XamlBuilder::WireButtonClickToAction(
+        IButton* button,
+        IAdaptiveActionElement* action,
+        IAdaptiveRenderContext* renderContext)
```

Comment:
Do you see a future where this would need the whole context rather than just the ActionInvoker?

Reply:
The reason I had this method take the RenderContext instead of ActionInvoker is because otherwise all callers of this method have to first get the ActionInvoker, which would be 2 extra lines of redundant code assuming callers already have the RenderContext.

Right now the only caller of this method is the WrapInFullWidthTouchTarget, but in the future we can likely make the action (button) rendering code also use this method. And when we do that, the button rendering code similarly doesn't have to write those 2 redundant lines of code to obtain the ActionInvoker before calling this method.

Reply:
Fair enough. The click handler is only taking the ActionInvoker in the lambda, so I'm good with this!

Reply:
Note also that I don't hold a reference to the Context. In the button click handler, I only pass along the actionInvoker, so there's no concern about a reference to the Context still being held after the method returns.